### PR TITLE
Fix newline character in GCC package

### LINKED ISSUE #4175: gcc or binutils seems to have been broken recently
My daily build two nights ago worked.  A fresh clone of github.com/llnl/spack late yesterday afternoon failed to build, as did my build last night.

I'm building gcc@5.4.0 with the CentOS system compiler, gcc@4.8.[35].

The compiler appears to be failing.

Here's a bit from a `config.log` file.  The interesting bit is `n--eh-frame-hdr`, it looks as if there might be a stray 'n' somewhere that's getting pasted onto the front of the `--eh-frame-hdr` switch:

```
configure:3851: /scratch/hartzelg/working-dir/workspace/daily-build/spack/lib/spack/env/gcc/gcc    conftest.c  >&5
/blah/apps/spack/v0.0.10/opt/spack/linux-centos7-x86_64/gcc-4.8.5/binutils-2.28-66ufycihw5nglnjuy4xf6hqpwrru6bx4/bin/ld: cannot find n--eh-frame-hdr: No
 such file or directory
collect2: error: ld returned 1 exit status
```

I've been fishing for the change, but perhaps it will ring a bell.

### PR DESCRIPTION
**Pull request by adamjstewart**
@hartzell Can you see if this fixes #4175? At this point I don't even understand Python anymore. If someone can explain raw strings to me I would be eternally grateful.

### CODE REVIEW

**Comment by hartzell** [comment]
in progress...

**Comment by hartzell** [comment]
Well, to my outsiders eye, the raw strings section of the Python docs [here](https://docs.python.org/2.0/ref/strings.html) seems kind of insane (not to mention a "values->valid" typo).

I think all that it's saying is that in most strings, escaped characters are transliterated, so that the two character string "\a" becomes the single character string whose ascii hex value is 0x07.

r'\a', on the other hand, remains the two character string composed of a `\` and an `a`.

In Perl, those would be "\a" and '\a'.

In this case you're just trying to cram enough spare backslashes into the string so that you end up with a "\n" by the time various things interpret it.

Now, if someone has an oracle that can predict how many backslashes are needed to escape a thing in any particular situation, *I'd* be eternally greatful... ;)

(I suspect @adamjstewart is joking, but I'll consider this a test of *my* understanding of Python)

**Comment by adamjstewart** [comment]
Yeah, that was my understanding too. Guess I got the wrong number of slashes?

**Comment by hartzell** [comment]
That didn't work.  I'm testing this:

```diff
diff --git a/var/spack/repos/builtin/packages/gcc/package.py b/var/spack/repos/builtin/packages/gcc/package.py
index d4ed234..41b6d2c 100644
--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -250,6 +250,6 @@ class Gcc(AutotoolsPackage):
             for line in lines:
                 out.write(line + '\n')
                 if line.startswith('*link:'):
-                    out.write(r'-rpath {0}:{1} \n'.format(
+                    out.write(r'-rpath {0}:{1} '.format(
                               self.prefix.lib, self.prefix.lib64))
         set_install_permissions(specs_file)
```

I can't see any non-asthetic [SIC?] to have a newline stuck in there....

**Comment by hartzell** [comment]
ps, the change in this PR ends up with 

```
6lar5z6rmmf5hrqme2ogbu4v/lib64 \\\n%{!static:--eh-frame-hdr}
```

in the string in `opt/spack/linux-centos7-x86_64/gcc-4.8.5/gcc-5.4.0-mqlhbvmx6lar5z6rmmf5hrqme2ogbu4v/lib/gcc/x86_64-unknown-linux-gnu/5.4.0/specs`

**Comment by adamjstewart** [comment]
The problem with that is we are parsing an existing file and writing out each line, so if you don't have the newline in there it will have 2 lines concatenated.

**Comment by hartzell** [comment]
Uhm, looking at the `specs` file, I think that might actually be what we want, all of this should be on the 

```
*link:
```

line.

**Comment by hartzell** [comment]
If this build fails I just go at the `specs` file with `vi` (or `emacs, if I get serious ;)) until I understand what needs to be there, then work back....

**Comment by adamjstewart** [comment]
Even with your change, it won't be on the link line. That line still gets a `\n`. It will be at the beginning of the line and concatenated with the following line. Which may or may not be correct. I don't know enough about it.

**Comment by hartzell** [comment]
This change Works For Me(tm):

```diff
diff --git a/var/spack/repos/builtin/packages/gcc/package.py b/var/spack/repos/builtin/packages/gcc/package.py
index d4ed234..41b6d2c 100644
--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -250,6 +250,6 @@ class Gcc(AutotoolsPackage):
             for line in lines:
                 out.write(line + '\n')
                 if line.startswith('*link:'):
-                    out.write(r'-rpath {0}:{1} \n'.format(
+                    out.write(r'-rpath {0}:{1} '.format(
                               self.prefix.lib, self.prefix.lib64))
         set_install_permissions(specs_file)
```

Here's the relevant section of the `.../lib/gcc/x86_64-unknown-linux-gnu/5.4.0/specs` file:

```
[...]
*endfile:
%{!mandroid|tno-android-ld:%{Ofast|ffast-math|funsafe-math-optimizations:crtfastmath.o%s}    %{mpc32:crtprec32.o%s}    %{mpc64:crtprec64.o%s}    %{mpc80:crtprec80.o%s} %{fvtable-verify=none:%s;      fvtable-verify=preinit:vtv_end_preinit.o%s;      fvtable-verify=std:vtv_end.o%s}    %{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s;:%{Ofast|ffast-math|funsafe-math-optimizations:crtfastmath.o%s}    %{mpc32:crtprec32.o%s}    %{mpc64:crtprec64.o%s}    %{mpc80:crtprec80.o%s} %{shared: crtend_so%O%s;: crtend_android%O%s}}

*link:
-rpath /scratch/hartzelg/spack-cime-sc1-working-dir/workspace/sc1-daily-build/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/gcc-5.4.0-mqlhbvmx6lar5z6rmmf5hrqme2ogbu4v/lib:/scratch/hartzelg/spack-cime-sc1-working-dir/workspace/sc1-daily-build/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/gcc-5.4.0-mqlhbvmx6lar5z6rmmf5hrqme2ogbu4v/lib64 %{!static:--eh-frame-hdr} %{!mandroid|tno-android-ld:%{m16|m32|mx32:;:-m elf_x86_64}                    %{m16|m32:-m elf_i386}                    %{mx32:-m elf32_x86_64}   %{shared:-shared}   %{!shared:     %{!static:       %{rdynamic:-export-dynamic}       %{m16|m32:-dynamic-linker %{muclibc:/lib/ld-uClibc.so.0;:%{mbionic:/system/bin/linker;:/lib/ld-linux.so.2}}}       %{m16|m32|mx32:;:-dynamic-linker %{muclibc:/lib/ld64-uClibc.so.0;:%{mbionic:/system/bin/linker64;:/lib64/ld-linux-x86-64.so.2}}}       %{mx32:-dynamic-linker %{muclibc:/lib/ldx32-uClibc.so.0;:%{mbionic:/system/bin/linkerx32;:/libx32/ld-linux-x32.so.2}}}}     %{static:-static}};:%{m16|m32|mx32:;:-m elf_x86_64}                    %{m16|m32:-m elf_i386}                    %{mx32:-m elf32_x86_64}   %{shared:-shared}   %{!shared:     %{!static:       %{rdynamic:-export-dynamic}       %{m16|m32:-dynamic-linker %{muclibc:/lib/ld-uClibc.so.0;:%{mbionic:/system/bin/linker;:/lib/ld-linux.so.2}}}       %{m16|m32|mx32:;:-dynamic-linker %{muclibc:/lib/ld64-uClibc.so.0;:%{mbionic:/system/bin/linker64;:/lib64/ld-linux-x86-64.so.2}}}       %{mx32:-dynamic-linker %{muclibc:/lib/ldx32-uClibc.so.0;:%{mbionic:/system/bin/linkerx32;:/libx32/ld-linux-x32.so.2}}}}     %{static:-static}} %{shared: -Bsymbolic}}

*lib:
%{!mandroid|tno-android-ld:%{pthread:-lpthread} %{shared:-lc}    %{!shared:%{mieee-fp:-lieee} %{profile:-lc_p}%{!profile:-lc}};:%{shared:-lc}    %{!shared:%{mieee-fp:-lieee} %{profile:-lc_p}%{!profile:-lc}} %{!static: -ldl}}
[...]
```

It looks "reasonable" (it *is* a gcc spec file, let's not get too hopeful...).  It's not "on the link line" (sorry, I spoke imprecisely...).  It's on the single line that goes with the '*link:' line.

I can use the resulting compiler to build things (e.g. built `jq` and it's prerequisites.

Convinced that this is the right change?

**Comment by adamjstewart** [comment]
@hartzell Yeah, that works for me. We can also drop the `r` now as there are no escape sequences to avoid. I'll update the PR.

**Comment by hartzell** [comment]
This version (without the raw string) also results in a working compiler for me.

**Comment by hartzell** [comment]
The travis failures look like unrelated failures.  Anything else to be done here?

**Comment by adamjstewart** [comment]
@hartzell See #4180. As soon as that PR is merged, I'll rebase this one.

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 1]
```diff
@@ -250,6 +250,6 @@ def write_rpath_specs(self):
             for line in lines:
                 out.write(line + '\n')
                 if line.startswith('*link:'):
-                    out.write(r'-rpath {0}:{1} \n'.format(
```
Comment (davydden) [inline var/spack/repos/builtin/packages/gcc/package.py]: the original line https://github.com/LLNL/spack/pull/4144/files#diff-38bf8d6101b9bc88270ac094f38565e1L186 was
```
out.write("-rpath %s/lib:%s/lib64 \\\n" %
                   (self.prefix, self.prefix))
```
so we are dropping that `\\\n` here. 

Not that i know the difference in Python between `\n` and `\\\n` :smile:

Comment (hartzell) [inline var/spack/repos/builtin/packages/gcc/package.py]: The spec line that resulted from that that version (e.g. fa4048aab6bb3ad3214daac069ea47fbe89109b9) looks like this:

```
*link:
-rpath /home/users/hartzelg/tmp/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/gcc-5.4.0-fyuuybt7ln3dgr5627p44lhnrztpwoqu/lib:/home/users/hartzelg/tmp/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/gcc-5.4.0-fyuuybt7ln3dgr5627p44lhnrztpwoqu/lib64 \
%{!static:--eh-frame-hdr} %{!mandroid|tno-android-ld:%{m16|m32|mx32:;:-m elf_x86_64}                    %{m16|m32:-m elf_i386}                    %{mx32:-m elf32_x86_64}   %{shared:-shared}   %{!shared:     %{!static:       %{rdynamic:-export-dynamic}       %{m16|m32:-dynamic-linker %{muclibc:/lib/ld-uClibc.so.0;:%{mbionic:/system/bin/linker;:/lib/ld-linux.so.2}}}       %{m16|m32|mx32:;:-dynamic-linker %{muclibc:/lib/ld64-uClibc.so.0;:%{mbionic:/system/bin/linker64;:/lib64/ld-linux-x86-64.so.2}}}       %{mx32:-dynamic-linker %{muclibc:/lib/ldx32-uClibc.so.0;:%{mbionic:/system/bin/linkerx32;:/libx32/ld-linux-x32.so.2}}}}     %{static:-static}};:%{m16|m32|mx32:;:-m elf_x86_64}                    %{m16|m32:-m elf_i386}                    %{mx32:-m elf32_x86_64}   %{shared:-shared}   %{!shared:     %{!static:       %{rdynamic:-export-dynamic}       %{m16|m32:-dynamic-linker %{muclibc:/lib/ld-uClibc.so.0;:%{mbionic:/system/bin/linker;:/lib/ld-linux.so.2}}}       %{m16|m32|mx32:;:-dynamic-linker %{muclibc:/lib/ld64-uClibc.so.0;:%{mbionic:/system/bin/linker64;:/lib64/ld-linux-x86-64.so.2}}}       %{mx32:-dynamic-linker %{muclibc:/lib/ldx32-uClibc.so.0;:%{mbionic:/system/bin/linkerx32;:/libx32/ld-linux-x32.so.2}}}}     %{static:-static}} %{shared: -Bsymbolic}}
```

That extra `\` is escaping the newline.  Not sure what that was considered useful, perhaps to make our addition stand out?

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gcc/package.py]: Oh, I see! So `'\\\n'` looks like `\` with a newline character after it and `r'\n'` just looks like `\n`. That actually makes sense. Thanks for teaching me about raw strings @hartzell!

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gcc/package.py]: I'm used to Expect, where you actually _do_ have to triply escape new line characters.

Comment (hartzell) [inline var/spack/repos/builtin/packages/gcc/package.py]: You're welcome.  That shaves epsilon off of the balance (thanks for all the help from your side)....

**Pull request merged**